### PR TITLE
Permute-Free Grouped GEMM for MoE (bf16, gfx950)

### PR DESCRIPTION
**Pull request by sudhu2k**
# Description

This PR adds a **permute-free** execution path for MoE `GroupedLinear` on ROCm. Instead of
physically permuting activations into expert-contiguous order before each grouped GEMM (and
un-permuting after), the gather is folded **into** the GEMM: the kernel reads each expert's rows
directly from the token-ordered activation buffer using a precomputed route list. This eliminates
the permute/unpermute activation copies, lets SwiGLU be treated as "just an op" (recomputed in the
backward rather than checkpointed), matches the MegaMOE reference kernel numerically, beats the
other bf16 grouped-GEMM backends currently in TE, and is fully **CUDA-graph capturable** (the
routing/align path is sync-free).

Opt-in via `NVTE_PERMUTE_FREE_GROUPED_GEMM=1` (ROCm builds only). ~6.9k LOC across 16 files; new
code is self-contained under `transformer_engine/pytorch/moe/` and
`transformer_engine/pytorch/flydsl_kernels/`, wired into `GroupedLinear`.

---

## 1. Motivation: what's expensive about the standard MoE path

A conventional MoE layer does, per micro-batch:

```
router → permute (scatter tokens into expert-contiguous order)
       → FC1 grouped GEMM  → SwiGLU activation
       → FC2 grouped GEMM
       → unpermute (scatter results back to token order) + combine
```

Two sources of avoidable **activation memory / bandwidth**:

1. **Permute/unpermute copies.** The permuted activation (`[num_routes, hidden]`) is a full extra
   materialization of the activations, written before FC1 and read/scattered after FC2 — plus the
   symmetric copies in the backward.
2. **SwiGLU checkpointing.** The standard path saves the `F`-wide activated tensor (and/or the `2F`
   pre-activation) for the backward, so the activation output lives across the whole fwd→bwd
   boundary.

For large expert counts and top-k this permute + activation traffic is a meaningful fraction of MoE
step time and peak memory.

---

## 2. Core idea: gather-in-GEMM over a route-list layout

Rather than moving data, we build a compact **route list** once and let the GEMM kernels do the
gather while they load their `A` tiles:

- The activation buffer stays in **received-token order** (`[num_recv_tokens, hidden]`), never
  permuted.
- A sync-free "align" pass builds an **expert-sorted, block-padded** index (`sorted_slot_ids`) that
  maps each padded GEMM slot → the received-token row to gather.
- The forward GEMM (`A = activations`, `B = [E, N, K]` weights) reads `A[sorted_slot_ids[slot], :]`
  inside the tile load — the gather is free (it rides the existing global loads), and no permuted
  copy is ever written.
- The token **combine** (and the FC1 input-grad reduction) is done by a **contention-free
  gather-combine** (each output token pulls its own routes) instead of an atomic scatter.

Because the whole routing/align path is over-allocated to static, shape-derived upper bounds and
carries the real extents as **device scalars** (no `.item()`, no host sync), the entire flow is
CUDA-graph capturable.

---

## 3. End-to-end flow

The same built routing metadata drives both FCs; a `route_space` flag selects direction (FC1 vs
FC2), so FC1 and FC2 share a single align build (`dataclasses.replace(meta, route_space=True)`).

```
                         received-token order                route/padded order
                         [num_recv, hidden]                  [em_max, ...]

FC1 (route_space=False)  x ──gather-in-GEMM (A gathered by sorted_slot_ids)──▶ [em_max, 2F]  (raw [gate|up])
                                                                                    │
SwiGLU (standalone op)   act(gate)*up*prob  ── recomputed, F-wide transient ◀───────┘  (only 2F preact persists)
                                                                                    │
FC2 (route_space=True)   [em_max, F] ──gather-in-GEMM (A read by route pos)──▶ [em_max, out]
                                                                                    │
combine                  gather-combine (each token pulls its routes) ──▶ [num_recv, out]
```

- **FC1 (`route_space=False`)** — input in token order. Forward *gathers* per expert
  (`index_a_by_route_pos=False`) into the padded `[em_max, 2F]` raw `[gate|up]` buffer. Dgrad
  combines the input gradient back to token rows via the contention-free gather-combine.
- **FC2 (`route_space=True`)** — input already in route order (FC1's output). The gated activation
  `act(gate)*up[*prob]` is applied in a **standalone pass** into an `F`-wide transient, then a plain
  gather-GEMM + gather-combine → `[num_recv, out]`. Dgrad gathers the token-space grad back into the
  compact route buffer.

---

## 4. Data structures

All routing state lives on `MoERoutingMetadata` / `PermuteFreeMetadata`
(`transformer_engine/pytorch/moe/moe_routing.py`). Everything below is built sync-free and
over-allocated to static bounds; real extents ride as device scalars.

| Field | Shape / type | Meaning |
|---|---|---|
| `routing_map` | `[num_recv_tokens, num_experts]` bool | The input: True where a token feeds a local expert. `num_routes = routing_map.sum()`. |
| `topk` | host int (optional) | Upper bound on experts a token can feed. Tightens the padded over-allocation from `T*E` down to `T*min(topk,E)` → smaller `em_max` (and less zero-init). |
| `sorted_slot_ids` | `[T*min(topk,E)]` i32 | **The route list.** Block-padded slot → received-token row to gather (sentinel `num_recv_tokens` for padding/tail → OOB read returns 0). |
| `expert_ids` | `[blocks_max]` i32 | Expert owning each `BLOCK_SIZE_M` block (`-1` past the real count; skipped). Drives the tile→expert map in the GEMM grid. |
| `slot_expert_ids` | `[T*min(topk,E)]` i32 | Per-slot expert id (`expert_ids[slot // block_size_m]`), for the standalone activation kernels. |
| `num_tokens_post_padded` | `[1]` i32 device scalar | Real padded extent `em`; bounds the kernels without a host sync. |
| `block_start` | `[num_experts]` i32 | Per-expert first block index (block units); expert `e`'s slots start at `block_start[e]*block_size_m`. (Consumed by backward wgrad to locate grad rows — see §6.) |
| `token_routes` / `token_route_count` | `[T, min(topk,E)]` i32 / `[T]` i32 | Token → its padded slot positions (inverse map). Powers the contention-free gather-combine that replaces the atomic scatter. |
| `block_size_m` | host int | `BLOCK_SIZE_M` the fwd/dgrad layout is padded to. |
| `route_counts` / `route_within` | cached tensors | Block-size-independent scan (per-expert counts + within-expert ranks) shared by the fwd/dgrad and wgrad align builds — computed once per routing map. |
| `wgrad_align` | `WgradAlign` (shared holder) | Separate block-`CONTRACT_M`(=32) align for the wgrad contraction (`sorted_slot_ids`, `block_start`, `blocks_per_expert`). Shared by reference across the FC1 metadata and its FC2 `replace` copy so the wgrad align is built once. |

Direction tag (`PermuteFreeMetadata`): `route_space` (FC1 vs FC2) and an optional `activation` hint
(`"silu"`/`"gelu"`) consumed on the FC2 direction. Per-route gating probabilities are **not** stored
here — they need a gradient, so they flow as a separate autograd tensor.

---

## 5. Kernels

**Routing / metadata (Triton, `pf_helper_kernels.py`)**
- `route_list_scan` — block-size-independent per-expert counts + within-expert ranks (shared across
  FC1/FC2/wgrad).
- `route_list_align` — builds `sorted_slot_ids`, `expert_ids`, `block_start`, `blocks_per_expert`,
  and the `token_routes` inverse map. Fully sync-free / graph-safe.
- `route_gather_combine` — contention-free per-token reduction over a token's routes (replaces
  atomic scatter for the FC2 combine and FC1 dgrad reduction).
- `fused_gated_act_prob_fwd` / `fused_gated_act_prob_bwd` — SwiGLU/GELU recompute + backward (see
  §6). `exp2`-based silu/gelu, no libdevice dependency.

**Grouped GEMMs (FlyDSL, `flydsl_kernels/permute_free_grouped_gemm/`)**
- `pf_fwd.py` — gather-in-GEMM forward. Tile-centric grid: `TILE_TO_GROUP` (= `expert_ids`) maps
  each M-tile to its expert; `A` rows are gathered by `sorted_slot_ids` (FC1) or read by route
  position (FC2). MegaMOE-ported fixed tile geometry (`32×32×16` MFMA, `BLOCK_N=256`, `GROUP_M=4`,
  XCD swizzle over the real tile range).
- `pf_dgrad.py` — data-gradient GEMM (NN contract against the transposed-weight view).
- `pf_wgrad.py` — weight-gradient GEMM. Expert-centric grid `N×K×E`; contracts over the **token**
  axis using a coalesced LDS fill + hardware transpose-read (`ds_read_tr16`) so the token slot
  becomes the matrix-core contraction axis with no strided global gather. Each `dW` tile is owned by
  exactly one workgroup → race-free store; supports bf16 or **fp32** output for in-kernel
  `main_grad` accumulation, and a `swap_gather` variant for FC2's native `dW` layout.

Thin dispatch/plumbing: `pf_fwd_wrapper.py`, `pf_wgrad_wrapper.py`, `tensor_shim.py`; orchestration
in `permute_free_grouped_gemm.py`.

---

## 6. Why this saves memory (the headline benefits)

**Removes the permute activation memory.** The activations are never copied into expert-contiguous
order. The GEMM gathers rows on the fly via `sorted_slot_ids`, so the `[num_routes, hidden]` permute
buffer (and its unpermute twin) simply doesn't exist — in both forward and backward. The token
combine is a gather (each token pulls its routes) rather than a scatter, so no atomic-scatter
scratch either.

**Removes the persisted SwiGLU activation.** FC1 emits the raw `2F` `[gate|up]` pre-activation; the
gated activation `act(gate)*up*prob` is materialized into a **short-lived `F`-wide transient** that
FC2 consumes immediately and is never saved. Across the fwd→bwd boundary we persist **only the `2F`
pre-activation** — the `F`-wide activated tensor that the standard path checkpoints is gone.

**Fuses the recompute into the SwiGLU backward — "swiglu is just an op."** In the backward,
`fused_gated_act_prob_fwd` re-materializes the `F`-wide activation from the saved `2F` preact, and
`fused_gated_act_prob_bwd` computes the grad w.r.t. the preact (and optionally the route-prob grad)
in the same pass. The recompute feeds the **unchanged** FC2 wgrad, so that kernel stays at full
stored-activation speed while only the `2F` preact is kept live. The whole SwiGLU becomes a single
fused op with no separately-checkpointed subgraph.

**Parity with the MegaMOE GEMM kernel.** The PF fwd GEMM kernels are ported from MegaMOE GEMM kernels, while wgrad is a custom written kernel specifically for Permute free.

**Faster than the other TE bf16 grouped-GEMM backends.**
`benchmarks/microbenchmarks/benchmark_perm_free_grouped_gemm.py` compares `permute_free` against
`hipblaslt`, `ck`, and `triton` on the DeepSeek/Grok/Qwen-style shapes; the permute-free path leads
for bf16 grouped GEMM.

**CUDA-graph capturable.** The align/routing pass has no host sync — it over-allocates to static,
shape-derived bounds and carries real extents as device scalars — and the GEMMs use fixed geometry
(no first-call autotune benchmarking). The full MoE step can therefore be captured into a HIP/CUDA
graph.

---

## 7. Integration & enablement

- Wired into `transformer_engine/pytorch/module/grouped_linear.py` (forward + backward dispatch,
  activation/route-prob handling).
- Public surface exported from `transformer_engine/pytorch/moe/__init__.py`.
- **Enable:** `NVTE_PERMUTE_FREE_GROUPED_GEMM=1` (ROCm/HIP builds only; gated by `IS_HIP_EXTENSION`).
  Off by default — zero impact on existing paths when unset.

---

## 8. Testing

- `tests/pytorch/test_perm_free_grouped_linear.py` (22 tests): FC1/FC2 fwd, dgrad, wgrad (incl. fp32
  accumulate and recompute-from-preact), the standalone gated-act fwd/bwd for both silu and gelu, the
  full FC1→FC2 gated pipeline, fwd/dgrad/wgrad consistency, gather-combine dispatch, sync-free align,
  and the enable flag.
- Verified on gfx950 (ROCm 7.2, torch `2.8.0+rocm7.2.1`): **22 passed**.
---

### Suggested reading order
1. `moe_routing.py` — the metadata contract and layout (start here).
2. `permute_free_grouped_gemm.py` module docstring + `permute_free_grouped_gemm_forward` /
   `_backward` — the dispatch/flow.
3. `pf_helper_kernels.py` — align, gated-act recompute/bwd, gather-combine.
4. `flydsl_kernels/permute_free_grouped_gemm/pf_fwd.py`, then `pf_dgrad.py`, then `pf_wgrad.py`.
5. `grouped_linear.py` diff — the integration seam.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/flydsl_kernels/gemm/gemm_common_utils.py [hunk 1]
```diff
@@ -220,3 +221,24 @@ def compute_global_swizzle(lane_id, wave_id, row_stride, n_rounds, preshuffled=F
...
+        if agpr_alloc < 0:
+            alloc = f"0,{-agpr_alloc}"
+        else:
+            alloc = f"{agpr_alloc},{agpr_alloc}"
+        attrs["passthrough"] = [
+            ["amdgpu-agpr-alloc", alloc],
+            ["amdgpu-mfma-vgpr-form", "false"],
+        ]
+    return attrs
+
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/flydsl_kernels/gemm/gemm_common_utils.py]: Fixed

File: benchmarks/microbenchmarks/benchmark_grouped_gemm.py [hunk 2]
```diff
@@ -89,6 +89,13 @@ def generate_deepseekv3_test_cases():
     )
 
 
+def generate_qwen3_235b_test_cases():
+    # Qwen3-235B-A22B: 128 routed experts, top-8, moe_intermediate_size 1536, hidden 4096.
+    return _generate_moe_test_cases(
+        "Qwen3-235B", n_routed_experts=128, moe_intermediate_size=1536, hidden_size=4096
+    )
+
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply benchmarks/microbenchmarks/benchmark_grouped_gemm.py:97]: Fixed.

File: transformer_engine/pytorch/flydsl_kernels/tensor_shim.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# SPDX-License-Identifier: MIT
+# Copyright (C) 2024-2026, Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/flydsl_kernels/tensor_shim.py]: Fixed.

File: transformer_engine/pytorch/moe/pf_helper_kernels.py [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+            (32, 256, 8),
+            (32, 128, 4),
+            (8, 256, 4),
+            (4, 512, 4),
+            (64, 128, 8),
+        )
+    ]
+
+
+@triton.autotune(configs=_gated_act_bwd_autotune_configs(), key=["F", "HAS_PROBS"])
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/pf_helper_kernels.py:143]: The standard capture pattern runs eager warmup iterations before capture, and the autotune caches are keyed coarsely (FlyDSL wgrad on (x.shape, N, num_experts, variant), Triton on (F, HAS_PROBS)), so every shape/variant the captured region launches is already benchmarked and cached by the time capture starts on a later iter. This change isn't needed.

File: tests/pytorch/test_perm_free_grouped_linear.py [hunk 5]
```diff
@@ -0,0 +1,881 @@
...
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+def _assert_close(actual: torch.Tensor, ref: torch.Tensor, dtype=torch.bfloat16) -> None:
+    tols = dtype_tols(dtype)
+    # Gather-combine (and similar reductions) can cancel to near-zero; those
+    # entries need an absolute floor. Use the same relative bound ``dtype_tols``
+    # already provides, scaled by peak |ref|.
+    tols["atol"] = max(tols["atol"], tols["rtol"] * ref.detach().float().abs().max().item())
+    assert_close(actual, ref, **tols)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply tests/pytorch/test_perm_free_grouped_linear.py]: Started using RMS instead of max for rtol scaling.

File: tests/pytorch/test_perm_free_grouped_linear.py [hunk 6]
```diff
@@ -0,0 +1,881 @@
...
+    gen = torch.Generator(device=device).manual_seed(seed)
+    routing_map = torch.zeros(num_recv_tokens, num_experts, dtype=torch.bool, device=device)
+    for t in range(num_recv_tokens):
+        k = int(torch.randint(1, max_hits + 1, (1,), device=device, generator=gen).item())
+        experts = torch.randperm(num_experts, device=device, generator=gen)[:k]
+        routing_map[t, experts] = True
+    # Guarantee every expert owns at least one route so per-expert refs are exercised.
+    for e in range(num_experts):
+        if not routing_map[:, e].any():
+            routing_map[int(e) % num_recv_tokens, e] = True
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply tests/pytorch/test_perm_free_grouped_linear.py:108]: Added tests with 0 routing experts and single expert edge case.

File: tests/pytorch/test_perm_free_grouped_linear.py [hunk 7]
```diff
@@ -0,0 +1,881 @@
...
+        # overwriting or accumulating in-kernel.
+        out = torch.zeros(weights_shape, device="cuda", dtype=torch.float32)
+        permute_free_grouped_gemm_bf16_fc2_wgrad(
+            fc2_input, grad_output, weights_shape, routing, out=out, accumulate=False
+        )
+        _assert_close(out, ref)
+        permute_free_grouped_gemm_bf16_fc2_wgrad(
+            fc2_input, grad_output, weights_shape, routing, out=out, accumulate=True
+        )
+        _assert_close(out, 2.0 * ref)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply tests/pytorch/test_perm_free_grouped_linear.py]: Fixed.

File: transformer_engine/pytorch/flydsl_kernels/gemm/pf_gemm_utils.py [hunk 8]
```diff
@@ -0,0 +1,1015 @@
...
+    wave_m_offset = wave_m * (N_TILES_A * 32)
+    base_row = block_m * BLOCK_M + wave_m_offset
+    base_col = block_n * BLOCK_N + wave_n_offset
+    store_c.store(c00_frag, base_row + 0, base_col + 0)
+    store_c.store(c01_frag, base_row + 0, base_col + LDS_BLOCK_N)
+    store_c.store(c10_frag, base_row + LDS_BLOCK_M, base_col + 0)
+    store_c.store(c11_frag, base_row + LDS_BLOCK_M, base_col + LDS_BLOCK_N)
+
+
+def gemm_bf16_nt_tile(
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/flydsl_kernels/gemm/pf_gemm_utils.py]: Refactored gemm_bf16_nn_tile.

File: transformer_engine/pytorch/module/grouped_linear.py [hunk 9]
```diff
@@ -1511,6 +1741,7 @@ def backward(
             weights = saved_tensors[num_inputs : num_inputs + N]
             saved_weights = saved_tensors[num_inputs + N : num_inputs + 2 * N]
             biases = saved_tensors[num_inputs + 2 * N : num_inputs + 3 * N]
+            dispatched_probs = saved_tensors[num_inputs + 3 * N]
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/module/grouped_linear.py]: Fixed.

File: transformer_engine/pytorch/module/grouped_linear.py [hunk 10]
```diff
@@ -869,17 +852,230 @@ def grouped_gemm_wgrad(*_unused):
...
+        if cpu_offloading:
+            mark_not_offload(*weights_fp8, *weights)
+
+        # Opt-in exact sizing: with ``m_splits`` carrying tokens_per_expert, its sum is the
+        # route count, which sizes the block-padded buffers exactly instead of to the static
+        # T * min(topk, E) bound. Set before the align is built (lazily, inside the wrapper);
+        # a metadata whose align is already cached keeps the bound it was built with.
+        m_splits_list = m_splits.tolist() if torch.is_tensor(m_splits) else list(m_splits)
+        if is_permute_free_exact_routes_enabled() and routing_metadata.num_routes is None:
+            routing_metadata.num_routes = sum(m_splits_list)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/module/grouped_linear.py]: Fixed.

File: transformer_engine/pytorch/moe/pf_helper_kernels.py [hunk 11]
```diff
@@ -0,0 +1,865 @@
...
+        return
+    r_offs = pid * BLOCK_M + tl.arange(0, BLOCK_M)
+    r_mask = r_offs < num_routes_bound
+    token = tl.load(token_ptr + r_offs, mask=r_mask, other=num_recv_tokens).to(tl.int64)
+    valid = token < num_recv_tokens
+    expert = tl.load(expert_ptr + r_offs, mask=r_mask, other=0).to(tl.int64)
+    if HAS_PROBS:
+        prob = tl.load(
+            probs_ptr + token * stride_pm + expert * stride_pe, mask=valid, other=0.0
+        ).to(tl.float32)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/pf_helper_kernels.py:214]: [T, topk] is not the natural post-dispatch layout of Megatron, it uses [T, E_local] probs and idx after _indices_to_multihot call in Megatron, but I've added _validate_dispatched_probs to validate the shape.

File: transformer_engine/pytorch/moe/permute_free_grouped_gemm.py [hunk 12]
```diff
@@ -0,0 +1,1141 @@
...
+    return metadata.route_counts, metadata.route_within
+
+
+def prepare_moe_align(metadata: MoERoutingMetadata, block_m: int) -> MoERoutingMetadata:
+    """Build and cache the fwd/dgrad route-list align buffers on ``metadata`` (sync-free)."""
+    if (
+        metadata.sorted_slot_ids is not None
+        and metadata.block_size_m == block_m
+        and metadata.token_routes is not None
+    ):
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/permute_free_grouped_gemm.py:341]: Fixed.

File: transformer_engine/pytorch/moe/permute_free_grouped_gemm.py [hunk 13]
```diff
@@ -0,0 +1,1141 @@
...
+    if route_space:
+        # FC2: grad is token-space [num_recv, out]. dgrad gathers back to the route buffer,
+        # then the standalone gated-activation backward maps dL/dF -> dL/d(2F) (+ dprob).
+        recompute = fc2_activation is not None and hidden_states is not None
+        # When both grads are wanted with a gated activation, fuse the wgrad's activation
+        # recompute into the act-bwd (which is already streaming the 2F preact) so the wgrad
+        # can reuse the F-wide fc2_input instead of re-reading the 2F preact a second time.
+        fused_fc2_input = None
+        if requires_dgrad:
+            weights_stacked = _stack_expert_weights(weights)  # [E, H, F], zero-copy when grouped
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/permute_free_grouped_gemm.py:1160]: Fixed.

File: transformer_engine/pytorch/moe/permute_free_grouped_gemm.py [hunk 14]
```diff
@@ -0,0 +1,1141 @@
...
+    # The grad operand lives in the forward's block-padded [R_block] slot layout (its row for
+    # route (e, w) is block_start[e]*block_size_m + w). Ensure the forward align is present so we
+    # can pass that padded per-expert base to the kernel; routes are contiguous within an expert,
+    # so base + within-rank indexes the correct padded grad row.
+    if routing.block_start is None or routing.block_size_m is None:
+        stub_w = torch.empty(
+            num_experts, out_features, in_features,
+            device=hidden_states.device, dtype=torch.bfloat16,
+        )
+        _ensure_fwd_align(routing, hidden_states, stub_w)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/permute_free_grouped_gemm.py]: Fixed.

File: tests/pytorch/test_perm_free_grouped_linear.py [hunk 15]
```diff
@@ -0,0 +1,881 @@
...
+    _wgrad_contract_m,
+    moe_align_route_list,
+)
+from transformer_engine.pytorch.moe.pf_helper_kernels import route_list_scan
+from utils import assert_close, dtype_tols
+
+pytestmark = pytest.mark.skipif(
+    not (IS_HIP_EXTENSION and torch.cuda.is_available()),
+    reason="Permute-free grouped GEMM tests require ROCm and CUDA device.",
+)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply tests/pytorch/test_perm_free_grouped_linear.py]: Fixed.

File: transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_wgrad.py [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+        K: fx.Int32,
+        num_recv_tokens: fx.Int32,
+    ):
+        bf16 = T.bf16
+        c0 = arith.constant(0, index=True)
+
+        dW_rsrc = ptr_rsrc(dW)
+        x_rsrc = ptr_rsrc(X)
+        grad_rsrc = ptr_rsrc(GRAD)
+        sorted_rsrc = ptr_rsrc(SORTED)
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_wgrad.py]: Fixed.

File: transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_fwd.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+    assert A.dtype == torch.bfloat16 and weight.dtype == torch.bfloat16
+    assert output.dtype == torch.bfloat16
+    c_m = int(output.shape[0])
+    E, N, K = weight.shape
+    assert K == A.shape[1], f"weight K={K} != A K={A.shape[1]}"
+    assert output.shape[1] == N, f"output N={output.shape[1]} != weight N={N}"
+
+    A = A.contiguous()
+    a_elems = int(A.numel())
+    if a_elems >= (1 << 31):
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_fwd.py:226]: Fixed.

File: transformer_engine/pytorch/module/grouped_linear.py [hunk 18]
```diff
@@ -981,6 +1180,33 @@ def forward(
                 pow2_grad=blockwise_recipe.fp8_quant_bwd_grad.power_2_scale,
             )
 
+        # Permute-free grouped GEMM (ROCm FlyDSL) opt-in. Gathers inside GEMM and
+        # returns early, bypassing the default permute / split setup below.
+        if _GroupedLinear._is_permute_free_grouped_gemm_supported(
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/module/grouped_linear.py]: Fixed.

File: transformer_engine/pytorch/moe/permute_free_grouped_gemm.py [hunk 19]
```diff
@@ -0,0 +1,1141 @@
...
+) -> int:
+    """Return ``routing.block_size_m``, building fwd align buffers when missing."""
+    if routing.sorted_slot_ids is not None and routing.block_size_m is not None:
+        return int(routing.block_size_m)
+    block_m = _fwd_align_block_size_m(routing.num_recv_tokens)
+    prepare_moe_align(routing, block_m)
+    return block_m
+
+
+def is_permute_free_grouped_gemm_enabled() -> bool:
```
[the remark this answers is not in the document]
Comment (sudhu2k) [reply transformer_engine/pytorch/moe/permute_free_grouped_gemm.py:173]: Fixed

File: transformer_engine/pytorch/moe/pf_fwd_wrapper.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
+
+"""FlyDSL permute-free MoE route-list forward gather-GEMM op wrapper.
```
Comment (aris134) [inline transformer_engine/pytorch/moe/pf_fwd_wrapper.py:4]: MoEBlaze citation needed?

File: transformer_engine/pytorch/flydsl_kernels/gemm/pf_gemm_utils.py [hunk 21]
```diff
@@ -0,0 +1,769 @@
...
+
+
+def wait_barrier(count):
+    _llvm.inline_asm(
+        res=None,
+        operands_=[],
+        asm_string=f"s_waitcnt vmcnt({count})\ns_barrier",
+        constraints="",
+        has_side_effects=True,
+    )
```
Comment (aris134) [inline transformer_engine/pytorch/flydsl_kernels/gemm/pf_gemm_utils.py:352]: I'm not sure we need this since we already have barrier defined in gemm_common_utils.py. Please check whether you can use the existing helper

**Comment by aris134** [comment]
Update README with usage of new environment variables introduced for the permute free grouped gemm

File: transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_fwd.py [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+    BLOCK_M=256,
+    BLOCK_N=256,
+    GROUP_M=4,
+    num_xcd=1,
+    nt_vmcnt=3,
+    waves_per_eu=2,
+    agpr_alloc=0,
+    gather=True,
+):
+    """Host entry: grouped bf16 NT GEMM. With ``gather=True`` (FC1) ``C[pos] = A[SORTED[pos]]
```
Comment (aris134) [inline transformer_engine/pytorch/flydsl_kernels/permute_free_grouped_gemm/pf_fwd.py:189]: Seems to be TN by TE convention.

**Review by aris134 — CHANGES_REQUESTED** [review]
